src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs

```diff
@@ -545,15 +564,17 @@ public void SendMessages(IEnumerable<RequestMessage> messages, MessageEncoderSet
             Ensure.IsNotNull(messages, nameof(messages));
             ThrowIfDisposedOrNotOpen();
 
-            var helper = new SendMessagesHelper(this, messages, messageEncoderSettings);
+            var helper = new SendMessagesHelper(this, messages, _sendCompressorType, _compressorFactory, messageEncoderSettings);
             try
             {
                 helper.EncodingMessages();
-                using (var buffer = helper.EncodeMessages(cancellationToken))
+                var buffers = helper.EncodeMessages(cancellationToken);
+                using (var efficientMessageBuffer = buffers.Item1)
```

Comment:
the idea here was that we can cover message not only by compression. For example, it could be `encryption` or similar.
So, in this case, the real transform operation should be added only on the lowest-level step here. I agree that the call stack/hierarchy is not very big here, but still.

`efficient` means the message which is ready to be used in the next steps. In other words - prepared to next steps.

how about `effectiveBuffer`?

Reply:
effective is no better than efficient

Maybe this whole logic needs to be revisited